src/main/java/org/broadinstitute/hellbender/engine/spark/FindAssemblyRegionsSpark.java

```diff
@@ -0,0 +1,215 @@
+            final SAMFileHeader header,
+            final SAMSequenceDictionary sequenceDictionary,
+            final String referenceFileName,
+            final FeatureManager features,
+            final List<ShardBoundary> intervalShards,
+            final Broadcast<Supplier<AssemblyRegionEvaluator>> assemblyRegionEvaluatorSupplierBroadcast,
+            final AssemblyRegionReadShardArgumentCollection shardingArgs,
+            final AssemblyRegionArgumentCollection assemblyRegionArgs,
+            final boolean includeReadsWithDeletionsInIsActivePileups,
+            final boolean shuffle) {
+        JavaRDD<Shard<GATKRead>> shardedReads = SparkSharder.shard(ctx, reads, GATKRead.class, sequenceDictionary, intervalShards, shardingArgs.readShardSize, shuffle);
+        Broadcast<FeatureManager> bFeatureManager = features == null ? null : ctx.broadcast(features);
+
+        // 1. Calculate activity for each locus in the desired intervals, in parallel.
+        JavaRDD<ActivityProfileStateRange> activityProfileStates = shardedReads.mapPartitions(getActivityProfileStatesFunction(referenceFileName, bFeatureManager, header,
+                assemblyRegionEvaluatorSupplierBroadcast, assemblyRegionArgs, includeReadsWithDeletionsInIsActivePileups));
+
+        // 2. Group by contig. We need to do this so we can perform the band pass filter over the whole contig, so we
+        // produce assembly regions that are identical to those produced by AssemblyRegionWalker.
+        // This step requires a shuffle, but the amount of data in the ActivityProfileStateRange should be small, so it
+        // should not be prohibitive.
+        JavaPairRDD<String, Iterable<ActivityProfileStateRange>> contigToGroupedStates = activityProfileStates
+                .keyBy((Function<ActivityProfileStateRange, String>) range -> range.getContig())
+                .groupByKey();
+
+        // 3. Run the band pass filter to find AssemblyRegions. The filtering is fairly cheap, so should be fast
+        // even though it has to scan a whole contig. Note that we *don't* fill in reads here, since after we have found
+        // the assembly regions we want to do assembly using the full resources of the cluster. So if we have
+        // very small assembly region objects, then we can collect them on the driver (or repartition them)
+        // for redistribution across the cluster, at which points the reads can be filled in. (See next two steps.)
+        JavaRDD<ReadlessAssemblyRegion> readlessAssemblyRegions = contigToGroupedStates
+                .flatMap(getReadlessAssemblyRegionsFunction(header, assemblyRegionArgs, includeReadsWithDeletionsInIsActivePileups));
+
+        // 4. Pull the assembly region boundaries down to the driver, so we can fill in reads.
+        List<ShardBoundary> assemblyRegionBoundaries = readlessAssemblyRegions
+                .map((Function<ReadlessAssemblyRegion, ShardBoundary>) FindAssemblyRegionsSpark::toShardBoundary)
```

Comment:
Why do you need toShardBoundary here when a ReaderlessAssemblyRegion is a shard bounadry?

Reply:
It's a problem with generics... `List<A>` cannot be substituted by `List<B>` even if `B` is a subclass of `A`. There should be a way of improving the code in this case though.

Reply:
I fixed this by generalizing the type in `SparkSharder`.